docs/rfc/active/5818-serial-updates.md

```diff
@@ -0,0 +1,313 @@
+through locking, transactions or Cassandra LWTs. More on this later.
+
+### 1: Change Processor
+
+Check the allowed set of transitions between current and proposed aspect state. This would be achieved with 
+business-specific logic injected into the DataHub codebase via plugins. This feels like teh wrong approach.
+
+Good
+* Fast as all code is executed in GMS
+
+Bad
+* No guarantee of state between the read and write
+* No guarantee that the plugin code will run
+* Specific business logic getting tied up with GMS code
+
+[![](https://mermaid.ink/img/pako:eNqFks9qAjEQxl9lyFV9gT1Yii3SgyB43cuQfLqB3USTWUXEd-_sH7vYCs0lYeY333yT5GZsdDCFyTi1CBYfng-JmzKQLrYSE61qjyBDZDgvlsvZerMr6CsckOU9H2FHQMOa3UDYsfBO61HQGk_MU3aheK_Ff1Rmq4q1wTZFi5xj0n55i9R4Ebi30WItdHzEhlDv87nyp8kVeYJCFFDyh0oo7umX58EwNXylis8g2ys6qpAwSbwcN3cbBVzojJR9DBP_evSE3Nbj6KgzyPnMdR0v_4y06KtDHCuDm26vyw6P9ZA3c9PoRbF3-ty3jiyNVGhQmkKPDnvuTJgy3BVtj2oTn86rUVPsWV3NDbcSd9dgTSGpxQMav8xI3b8BFXTCxw)](https://mermaid.live/edit#pako:eNqFks9qAjEQxl9lyFV9gT1Yii3SgyB43cuQfLqB3USTWUXEd-_sH7vYCs0lYeY333yT5GZsdDCFyTi1CBYfng-JmzKQLrYSE61qjyBDZDgvlsvZerMr6CsckOU9H2FHQMOa3UDYsfBO61HQGk_MU3aheK_Ff1Rmq4q1wTZFi5xj0n55i9R4Ebi30WItdHzEhlDv87nyp8kVeYJCFFDyh0oo7umX58EwNXylis8g2ys6qpAwSbwcN3cbBVzojJR9DBP_evSE3Nbj6KgzyPnMdR0v_4y06KtDHCuDm26vyw6P9ZA3c9PoRbF3-ty3jiyNVGhQmkKPDnvuTJgy3BVtj2oTn86rUVPsWV3NDbcSd9dgTSGpxQMav8xI3b8BFXTCxw)
+
+### 2: Read and Conditional Update
+
+Read the current aspect and pass it back to the client. The client can propose an update on the basis that what they 
+just read is still valid.
+
+Good
+* Guarantee of state between the read and write
+* No need for business-specific logic in GMS code
+
+Bad
+* Slow, especially under high contention
+* Client will need to handle retries
+
+[![](https://mermaid.ink/img/pako:eNqVkb9qwzAQxl9FaG3yAhpcQhsylJBi083LIZ0bgSwp8mkoIe_ecyQXl2appuO733f_dJU6GJRKTnjJ6DW-WvhMMPZe8ANNIYkXZ9FTUVwIUeyIcIwksifrBMTIgBFMYkohFbCYtk3zdDh2ShyQdlNEXcuwtuVcgZTQOSUO1kT1r6B3IH0W8JdZenxEA4SliLCDmHg696j23L1pjkjABuh4R_yH-5dvHvDevF7huR7O0aIU4cHSp7eSQjfh-nIP0H3bntpKe_MTyI0cMY1gDX_gdZZ7SWccsZeKQ4MDZEe97P2NUcgUui-vpaKUcSPzfeH631INwHMs6t5Y3q6Kt29ZMrRr)](https://mermaid.live/edit#pako:eNqVkb9qwzAQxl9FaG3yAhpcQhsylJBi083LIZ0bgSwp8mkoIe_ecyQXl2appuO733f_dJU6GJRKTnjJ6DW-WvhMMPZe8ANNIYkXZ9FTUVwIUeyIcIwksifrBMTIgBFMYkohFbCYtk3zdDh2ShyQdlNEXcuwtuVcgZTQOSUO1kT1r6B3IH0W8JdZenxEA4SliLCDmHg696j23L1pjkjABuh4R_yH-5dvHvDevF7huR7O0aIU4cHSp7eSQjfh-nIP0H3bntpKe_MTyI0cMY1gDX_gdZZ7SWccsZeKQ4MDZEe97P2NUcgUui-vpaKUcSPzfeH631INwHMs6t5Y3q6Kt29ZMrRr)
+
+### 3: Idempotent Client Update
+
+The client proposes a partial change to an aspect that is idempotent. The change can apply whatever state the aspect 
+is in at the time. This will need to be extended to check for multiple currentAspect states when requested.
+
+Good
+* No need for business-specific logic in GMS code
+* Fast, as retry loop will be confined to GMS
+
+Bad
+* GMS will need to understand partial (PATCH) updates
```

Comment:
@RyanHolstien Please have a look here. 

With Ryan's changes, we'll be able to support a specific set of PATCH formats, but it may not always suffice for addressing the above use case (if we don't formally support a patch of the given type. Ryan can elaborate.

Reply:
As long as the path drills down far enough, JsonPatch semantics are capable of covering this scenario. Basically you can specify precise fields to modify and it will not impact the rest of the aspect, but if you target the top level of the aspect you may be overwriting things. Example:

JsonPatch for Ownership at top level:
op: ADD
path: /
value:
{ owners: { urn: urn:li:corpuser:owner1 : {...} }

vs JsonPatch at lower level:
op: ADD
path: /owners/urn:li:corpuser:owner1
value: {...}

The first one will wipe out other values in the array, but the second one will not, this is why the SDK implementation is very limited in scope since JsonPatch is almost _too_ flexible in what it allows. It can very much accommodate this use case though.

Reply:
Makes sense, thanks. I'd be keen to know if JsonPatch can handle idempotent updates like "ensure this set contains these values". But I guess clients will need some form of serialisation guarantees regardless of whether it's a PATCH update or a full replacement.

Reply:
Those are two different things in my mind:

1) Idempotency of updates -> Guarantees that if you add owner1 multiple times, it doesn't add it repeatedly. This is ensured

2) Verification of particular values -> Guarantees that owner1 is only added if owner2 is present (or something to this effect? Not sure exactly what you mean here). This is not ensured. Maybe you mean:

add owner1 happens at same time as add owner2, owner1 AND owner2 should appear?

This should also be ensured per transactional consistency.